tree_sitter/binding.c

```diff
@@ -1939,6 +1968,90 @@ static bool satisfies_text_predicates(Query *query, TSQueryMatch match, Tree *tr
     return false;
 }
 
+static PyObject *query_matches(Query *self, PyObject *args, PyObject *kwargs) {
+    ModuleState *state = PyType_GetModuleState(Py_TYPE(self));
+    char *keywords[] = {
+        "node", "start_point", "end_point", "start_byte", "end_byte", NULL,
+    };
+
+    Node *node = NULL;
+    TSPoint start_point = {.row = 0, .column = 0};
+    TSPoint end_point = {.row = UINT32_MAX, .column = UINT32_MAX};
+    unsigned start_byte = 0, end_byte = UINT32_MAX;
+
+    int ok = PyArg_ParseTupleAndKeywords(args, kwargs, "O|(II)(II)II", keywords, (PyObject **)&node,
+                                         &start_point.row, &start_point.column, &end_point.row,
+                                         &end_point.column, &start_byte, &end_byte);
+    if (!ok) {
+        return NULL;
+    }
+
+    if (!PyObject_IsInstance((PyObject *)node, (PyObject *)state->node_type)) {
+        PyErr_SetString(PyExc_TypeError, "First argument to captures must be a Node");
+        return NULL;
+    }
+
+    ts_query_cursor_set_byte_range(state->query_cursor, start_byte, end_byte);
+    ts_query_cursor_set_point_range(state->query_cursor, start_point, end_point);
+    ts_query_cursor_exec(state->query_cursor, self->query, node->node);
+
+    QueryMatch *match = NULL;
+    PyObject *result = PyList_New(0);
+    if (result == NULL) {
+        goto error;
+    }
+    PyObject *captures_for_match = PyList_New(0);
+
+    TSQueryMatch _match;
+    while (ts_query_cursor_next_match(state->query_cursor, &_match)) {
+        match = (QueryMatch *)query_match_new_internal(state, _match);
+        if (match == NULL) {
+            goto error;
+        }
+        PyObject *captures_for_match = PyList_New(0);
+        if (captures_for_match == NULL) {
+            goto error;
+        }
+        for (unsigned i = 0; i < _match.capture_count; i++) {
+            QueryCapture *capture =
+                (QueryCapture *)query_capture_new_internal(state, _match.captures[i]);
+            if (capture == NULL) {
+                Py_XDECREF(captures_for_match);
+                goto error;
+            }
+            if (satisfies_text_predicates(self, _match, (Tree *)node->tree)) {
```

Comment:
Could this check be moved outside of the capture loop to line 2011? The result of satisfies_text_predicates() should be the same the for all captures in the match so it could be checked just once per match.